src/include/sof/audio/buffer.h

```diff
@@ -276,4 +276,23 @@ static inline int buffer_set_params(struct comp_buffer *buffer,
 	return 0;
 }
 
+/*
+ * This should be used for pipeline params negotiation only,
+ * where the component has more than 1 input or output, and
+ * the component itself only doing channel muxing/demuxing
+ * only so no frame_fmt or rate varies among buffers connected
+ * to it.
+ */
+static inline bool buffer_params_match(struct comp_buffer *buffer,
+				       struct sof_ipc_stream_params *params)
+{
+	if (!params || !buffer) {
+		trace_buffer_error("buffer_params_match(): invalid");
```

Comment:
Trace message does not explain what happened. I am also not sure whether this is real run-time error, expected scenario (the message should be a warning at most then), or a logic error (and assert() should be used here)?

Reply:
@mmaka1 Asserts cannot be used as _sole_ verification, because they can be disabled. Not yet, but once we will move to Zephyr we will be able to disable asserts. So if it's a logic error, we shouldn't have _just_ assert but also this trace and return.

And actually... We shouldn't assert on parameters that are controlled by the Linux driver, but only internal values. Asserting on an external guarantee (something the Linux driver "guarantees" is never a good idea, because mismatched versions will occur).

Reply:
@paulstelian97 I assume that asserts may be compiled out of the production code, so they should be use to catch logic errors only (those coming from coding issues) during the code bringup and debug phase. Any data coming from the driver should be verified at run-time. However I asked about this particular case, afair the parameters in this flow are internally computed by the fw, so I asked to double check whether tracing run-time error makes sense in this case. Still not sure what response is expected when seeing this error message in the trace log.

Reply:
If we don't know for certain this is internally computed, assume it isn't. That's the thing.

Reply:
@mmaka1 @paulstelian97 Thanks for the comments.

We are not sure how the helper will be used and what params and buffer it could be passed in, so I treated it as run-time error here.

But after thought about it more now, as @mmaka1 stated, here what we tried to catch here is the code logic error(e.g. some calling to buffer_params_match() with NULL buffer or params) but not the run-time params errors from the Linux kernel side, the response should be panicking the FW but not just returning an false when this error happen, so changing it to 'assert' handling looks good to me.